File: rust/src/src/app_helpers.rs
// src/app_helpers.rs  <- src/app_helpers.py
//! Small filesystem / encoding helpers used across the app.

use crate::pyos as os;
use base64::Engine as _;
use std::ffi::OsString;
use std::path::{Component, Path, PathBuf};

/// Read file if it exists, return empty string otherwise.
pub fn read_if_exists(path: &str) -> String {
    // try: with open(path, "r", encoding="utf-8") as f: return f.read().strip()
    // except Exception: return ""
    match std::fs::read_to_string(path) {
        // Python text mode performs universal-newline translation: every
        // "\r\n" and bare "\r" becomes "\n" before .strip(). Rust's
        // read_to_string returns raw bytes, so normalize to match.
        Ok(s) => s.replace("\r\n", "\n").replace('\r', "\n").trim().to_string(),
        Err(_) => String::new(),
    }
}

/// Convert file to data URL.
///
/// Mirrors `open(path, "rb")` then base64-encoding the bytes; if the file
/// cannot be read the underlying `OSError` is propagated (Python would raise).
pub fn file_to_data_url(path: &str, mime: &str) -> crate::error::PyResult<String> {
    // with open(path, "rb") as f: b64 = base64.b64encode(f.read()).decode("ascii")
    let data = std::fs::read(path)?;
    let b64 = base64::engine::general_purpose::STANDARD.encode(&data);
    // return f"data:{mime};base64,{b64}"
    Ok(format!("data:{mime};base64,{b64}"))
}

/// Join paths and return absolute path.
pub fn abs_join(base_dir: &str, rel: &str) -> String {
    // return os.path.abspath(os.path.join(base_dir, rel))
    abspath(&os::path::join(base_dir, rel))
}

/// Check if path is inside base directory.
pub fn inside_base_dir(base_dir: &str, path: &str) -> bool {
    // base = os.path.realpath(base_dir)
    // p = os.path.realpath(path)
    let base = realpath(base_dir);
    let p = realpath(path);
    // try: return os.path.commonpath([base, p]) == base
    // except Exception: return False
    match commonpath(&[&base, &p]) {
        Some(common) => common == base,
        None => false,
    }
}

// --- helpers mirroring os.path semantics ------------------------------------

/// `os.path.abspath`: normalize and make absolute against the current working
/// directory, without touching the filesystem (no symlink resolution).
fn abspath(path: &str) -> String {
    let p = Path::new(path);
    let joined: PathBuf = if p.is_absolute() {
        p.to_path_buf()
    } else {
        let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
        cwd.join(p)
    };
    normpath(&joined)
}

/// `os.path.realpath`: resolve symlinks on the longest *existing* ancestor of
/// the path and lexically re-append the nonexistent tail, mirroring CPython.
///
/// `std::fs::canonicalize` alone errors on any nonexistent component, which
/// would silently drop us to a non-symlink-resolving `abspath` — wrong on
/// systems where parents are symlinks (macOS `/tmp`, `/var`, `/etc`). Since
/// `inside_base_dir` is a security boundary, the two operands must resolve
/// against the same root, so we resolve the existing prefix explicitly.
pub(crate) fn realpath(path: &str) -> String {
    let abs = abspath(path); // absolute + lexically normalized (no `.`/`..`)
    let mut prefix = PathBuf::from(&abs);
    // Components stripped off the (nonexistent) tail, deepest first.
    let mut tail: Vec<OsString> = Vec::new();
    loop {
        match std::fs::canonicalize(&prefix) {
            Ok(mut real) => {
                // Re-append the tail in original (top-down) order.
                for comp in tail.iter().rev() {
                    real.push(comp);
                }
                return normpath(&real);
            }
            Err(_) => match prefix.file_name() {
                Some(name) => {
                    tail.push(name.to_os_string());
                    if !prefix.pop() {
                        break;
                    }
                }
                None => break,
            },
        }
    }
    // Even the root could not be canonicalized — fall back to the lexical form.
    abs
}

/// Lexical path normalization (collapse `.`/`..`, dedup separators), matching
/// `os.path.normpath` closely enough for the comparisons used here.
fn normpath(path: &Path) -> String {
    let mut out: Vec<Component> = Vec::new();
    for comp in path.components() {
        match comp {
            Component::CurDir => {}
            Component::ParentDir => {
                match out.last() {
                    Some(Component::Normal(_)) => {
                        out.pop();
                    }
                    Some(Component::RootDir) | Some(Component::Prefix(_)) => {}
                    _ => out.push(comp),
                }
            }
            other => out.push(other),
        }
    }
    let mut result = PathBuf::new();
    for comp in out {
        result.push(comp.as_os_str());
    }
    let s = result.to_string_lossy().into_owned();
    if s.is_empty() {
        ".".to_string()
    } else {
        s
    }
}

/// `os.path.commonpath`: longest common sub-path. Returns `None` (the Python
/// `ValueError`) when the inputs mix absolute and relative paths.
pub(crate) fn commonpath(paths: &[&str]) -> Option<String> {
    if paths.is_empty() {
        return None;
    }
    let split = |p: &str| -> Vec<String> {
        Path::new(p)
            .components()
            .map(|c| c.as_os_str().to_string_lossy().into_owned())
            .collect()
    };
    let first_abs = Path::new(paths[0]).is_absolute();
    for p in paths {
        if Path::new(p).is_absolute() != first_abs {
            return None; // raise ValueError: Can't mix absolute and relative paths
        }
    }
    let split_paths: Vec<Vec<String>> = paths.iter().map(|p| split(p)).collect();
    let min_len = split_paths.iter().map(|v| v.len()).min().unwrap_or(0);
    let mut common: Vec<String> = Vec::new();
    for i in 0..min_len {
        let seg = &split_paths[0][i];
        if split_paths.iter().all(|v| &v[i] == seg) {
            common.push(seg.clone());
        } else {
            break;
        }
    }
    let mut result = PathBuf::new();
    for seg in &common {
        result.push(seg);
    }
    Some(result.to_string_lossy().into_owned())
}

File: src/app_helpers.py
# src/app_helpers.py
import base64
import json
import logging
import os

from fastapi import HTTPException
from fastapi.responses import HTMLResponse
from starlette.requests import Request

logger = logging.getLogger(__name__)


def _login_theme_json() -> str:
    """Active profile theme as a script-safe JSON literal for the login page.

    The pre-auth login page can't fetch /api/prefs/theme (401) and a fresh
    device has nothing in localStorage, so without this it always renders the
    built-in default. We inject the profile's active theme so login matches
    the theme set in-app. Only a single-user instance is handled — with more
    than one user there's no way to know whose theme to show before sign-in,
    so we return "null" and let the client fall back to localStorage/default.
    Returns the string "null" (a valid JS literal) on any miss or error.
    """
    try:
        from src.constants import USER_PREFS_FILE
        with open(USER_PREFS_FILE, "r", encoding="utf-8") as f:
            users = (json.load(f) or {}).get("_users", {})
        if len(users) != 1:
            return "null"
        theme = next(iter(users.values())).get("theme")
        if not isinstance(theme, dict) or not theme.get("colors"):
            return "null"
        # Escape so the JSON can't break out of the surrounding <script>.
        return (
            json.dumps(theme)
            .replace("<", "\\u003c")
            .replace(">", "\\u003e")
            .replace("&", "\\u0026")
        )
    except Exception:
        return "null"

def read_if_exists(path: str) -> str:
    """Read file if it exists, return empty string otherwise."""
    try:
        with open(path, "r", encoding="utf-8") as f:
            return f.read().strip()
    except Exception:
        return ""

def file_to_data_url(path: str, mime: str) -> str:
    """Convert file to data URL."""
    with open(path, "rb") as f:
        b64 = base64.b64encode(f.read()).decode("ascii")
    return f"data:{mime};base64,{b64}"

def abs_join(base_dir: str, rel: str) -> str:
    """Join paths and return absolute path."""
    return os.path.abspath(os.path.join(base_dir, rel))

def serve_html_with_nonce(request: Request, file_path: str) -> HTMLResponse:
    """Read an app-bundled HTML page and inject the CSP nonce into inline <script> tags.

    Callers pass fixed, server-owned template paths (index/login/backgrounds),
    never a client-supplied path. So any read failure here — a missing file
    (broken deployment) or a permission/IO error — is a server fault, not a
    client "not found": map all of them to a logged 500 so a missing core
    template surfaces in 5xx alerting instead of hiding behind a 404. If a
    future caller serves a client-influenced path where 404 is correct, branch
    that at the call site rather than defaulting this shared helper to 404.
    """
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            html = f.read()
    except OSError:
        logger.exception("Failed to read page %s", file_path)
        raise HTTPException(500, "Internal server error")
    nonce = getattr(request.state, "csp_nonce", "")
    html = html.replace("{{CSP_NONCE}}", nonce)
    # Only the login page carries this placeholder; skip the prefs read for
    # every other template (index, backgrounds) that doesn't need it.
    if "{{LOGIN_THEME}}" in html:
        html = html.replace("{{LOGIN_THEME}}", _login_theme_json())
    return HTMLResponse(html)


def inside_base_dir(base_dir: str, path: str) -> bool:
    """Check if path is inside base directory."""
    if not isinstance(base_dir, str) or not isinstance(path, str):
        return False
    base = os.path.realpath(base_dir)
    p = os.path.realpath(path)
    try:
        return os.path.commonpath([base, p]) == base
    except Exception:
        return False
